alica_engine/src/engine/Assignment.cpp

```diff
@@ -1,532 +1,448 @@
-        rlist->push_back(r);
-        robotStateMapping->setState(r, defep->getState());
-    }
+    const State* s = _plan->getStateByID(sid);
+    return s ? AgentsInStateView(this, s) : AgentsInStateView();
 }
 
-bool Assignment::removeRobot(const supplementary::AgentID* robotId)
+void Assignment::clear()
 {
-    this->robotStateMapping->removeRobot(robotId);
-    AgentGrp* curRobots;
-    for (int i = 0; i < this->epRobotsMapping->getSize(); ++i) {
-        curRobots = this->epRobotsMapping->editRobots(i);
-        auto iter = find_if(curRobots->begin(), curRobots->end(), [&robotId](const supplementary::AgentID* id) { return *robotId == *id; });
-        if (iter != curRobots->end()) {
-            curRobots->erase(iter);
-            return true;
-        }
+    _successData.clear();
+    for (AgentStatePairs& asp : _assignmentData) {
+        asp.clear();
     }
-    return false;
 }
 
-void Assignment::addRobot(const supplementary::AgentID* id, const EntryPoint* e, const State* s)
+bool Assignment::updateAgent(AgentIDConstPtr agent, const EntryPoint* e)
 {
-    if (e == nullptr) {
-        return;
-    }
-    this->robotStateMapping->setState(id, s);
-    this->epRobotsMapping->addRobot(id, e);
-    return;
+    return updateAgent(agent, e, nullptr);
 }
 
-/**
- * Tests whether this assignment is valid with respect to the plan's cardinalities.
- * @return A bool
- */
-bool Assignment::isValid() const
+bool Assignment::updateAgent(AgentIDConstPtr agent, const EntryPoint* e, const State* s)
 {
-    const vector<shared_ptr<std::list<const supplementary::AgentID*>>>& success = this->epSucMapping->getRobots();
+    bool found = false;
+    bool inserted = false;
+    int i = 0;
+    assert(s == nullptr || s->getEntryPoint() == e);
 
-    for (int i = 0; i < this->epRobotsMapping->getSize(); ++i) {
-        int c = this->epRobotsMapping->getRobots(i)->size() + success[i]->size();
-        if (c > this->epRobotsMapping->getEp(i)->getMaxCardinality() || c < this->epRobotsMapping->getEp(i)->getMinCardinality()) {
-            return false;
+    for (AgentStatePairs& asps : _assignmentData) {
+        const bool isTargetEp = e == _plan->getEntryPoints()[i];
+        if (isTargetEp) {
+            for (AgentStatePair& asp : asps) {
+                if (asp.first == agent) {
+                    // assume a null state signals no change
+                    if (s == nullptr || asp.second == s) {
+                        return false;
+                    } else {
+                        asp.second = s;
+                        return true;
+                    }
+                }
+            }
+            asps.emplace_back(agent, s ? s : e->getState());
+            inserted = true;
+        } else if (!found) {
+            for (int j = 0; j < static_cast<int>(asps.size()); ++j) {
+                if (asps.getRaw()[j].first == agent) {
+                    found = true;
+                    asps.removeAt(j);
```

Comment:
why removed?

Reply:
not sure what you mean. The vector is not resized, but an element is removed.

Reply:
I can't find AgentStatePairs. on line 299 we have an emplace_back. I assumed that the removeAt and the emplace_back were complementary actions.